File: room/lib/room.ex
defmodule Room do
  alias Room.{Chat, Player}
  
  def find_or_create(name) do
    Room.DynamicSupervisor.find_or_create(name)
    name
  end

  def get(name) do
    %{chat: chat(name), player: player(name)}
  end

  def chat(name) do
    Chat.get(name)
  end
  
  def join(name) do
    Chat.join(name)
  end
  
  def add_message(name, user_id, body) do
    Chat.add_message(name, user_id, body)
  end

  # TODO: is there a better place to put this logic?
  def leave(name, user_id) do
    users_left = name
    |> Chat.leave(user_id)
    |> Chat.count_users()

    if users_left == 0 do
      Player.pause(name)
    end

    name
  end

  def player(name) do
    Player.get(name)
  end

  def play(name) do
    Player.play(name)
  end

  def pause(name) do
    Player.pause(name)
  end

  def seek(name, timestamp) do
    Player.seek(name, timestamp)
  end

  def new_video(name, url) do
    Player.new_video(name, url)
  end
end

File: client/src/Room.js
import React, { useEffect } from 'react';
import styled from 'styled-components';

import useChannel from './useChannel';

import VideoInput from './VideoInput';
import RoomStatus from './RoomStatus';
import Chat from './Chat';
import Player from './Player';

const Wrapper = styled.div`
  padding: 1rem;
  grid-gap: 1rem;
  display: grid;
  grid-template-columns: 1fr 400px;
  grid-template-rows: 4rem auto;
  width: 100%;
  height: 100%;
  position: absolute;
`;

export default ({ name }) => {
  const [push, select] = useChannel(name);

  useEffect(
    () => {
      document.title = `#${name}`;
    },
    [name],
  );

  if (!select.getIsConnected()) {
    return <div />;
  }

  return (
    <Wrapper>
      <VideoInput
        url={select.getUrl()}
        onNewVideo={url => push.newVideo({ url })}
      />
      <RoomStatus users={select.getUsers()} />
      <Player
        url={select.getUrl()}
        isPlaying={select.getIsPlaying()}
        timestamp={select.getTimestamp()}
        onPlay={push.play}
        onPause={push.pause}
        onSeek={timestamp => push.seek({ timestamp })}
      />
      <Chat
        messages={select.getMessages()}
        users={select.getUsers()}
        typingUsers={select.getTypingUsers()}
        onAddMessage={body => push.addMessage({ body })}
        onTyping={isTyping => push.userTyping({ isTyping })}
      />
    </Wrapper>
  );
};
